simple_certmanager/forms.py

```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
             instance.save()
         return instance
 
+    def clean(self) -> dict[str, Any]:
+        # Errors if the CSR already has a certificate
+        if self.instance.public_certificate:
+            self.add_error(
+                "certificate",
+                "A certificate already exists for this CSR. "
+                "Delete the certificate first.",
+            )
+        # Handle the certificate file upload
+        if "certificate" in self.cleaned_data and self.cleaned_data["certificate"]:
+            uploaded_file = self.cleaned_data["certificate"]
+            try:
+                uploaded_file.seek(0)
+                certificate_data = uploaded_file.read()
+
+                # Load the certificate and throw an error if the certificate is invalid
+                certificate = x509.load_pem_x509_certificate(certificate_data)
+                private_key = load_pem_x509_private_key(
+                    self.instance.private_key.encode()
+                )
+
+                # Check if the certificate matches the CSR by comparing the public keys
+                private_key_pubkey = private_key.public_key()
+                certificate_public_key = certificate.public_key()
+                match = private_key_pubkey == certificate_public_key
+                if not match:
+                    self.add_error(
+                        "certificate",
+                        "Certificate does not match the signature from the actual CSR.",
+                    )
+
+            except ValueError:
+                self.add_error(
+                    "certificate", "Invalid certificate. Check the file format."
```

Comment:
there are many lines in the `try` block that can raise a `ValueError`, e.g. loading a private key with password

Reading the file contents is also not something that should be inside this `try` block, we shouldn't expect that to error otherwise we need to file a bug report with Django.

Try to wrap only the code you expect to raise an exception, so rather do this instead:

```py
uploaded_file.seek(0)
certificate_data = uploaded_file.read()
try:
    certificate = x509.load_pem_x509_certificate(certificate_data)
except ValueError:
    ...

# not wrapped, we control the private key and failing to load it means a bug in
# our code that needs fixing, the end-user cannot solve any validation errors we
# show
private_key = load_pem_x509_private_key(self.instance.private_key.encode())

... # validate public keys against each other etc. etc.
```